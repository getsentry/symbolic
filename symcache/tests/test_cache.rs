use std::fmt;

use failure::Error;
use insta;

use symbolic_common::ByteView;
use symbolic_symcache::SymCache;

/// Helper to create neat snapshots for symbol tables.
struct FunctionsDebug<'a>(&'a SymCache<'a>);

impl fmt::Debug for FunctionsDebug<'_> {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        for result in self.0.functions() {
            match result {
                Ok(function) => writeln!(f, "{:>16x} {}", &function.address(), &function.name())?,
                Err(error) => writeln!(f, "{:?}", error)?,
            }
        }

        Ok(())
    }
}

#[test]
fn test_load_header_linux() -> Result<(), Error> {
    let buffer = ByteView::open("../testutils/fixtures/symcache/current/linux.symc")?;
    let symcache = SymCache::parse(&buffer)?;
    insta::assert_debug_snapshot_matches!(symcache, @r###"SymCache {
    debug_id: DebugId {
        uuid: "c0bcc3f1-9827-fe65-3058-404b2831d9e6",
        appendix: 0
    },
    arch: Amd64,
    has_line_info: true,
    has_file_info: true,
    functions: 1955
}"###);
    Ok(())
}

#[test]
fn test_load_functions_linux() -> Result<(), Error> {
    let buffer = ByteView::open("../testutils/fixtures/symcache/current/linux.symc")?;
    let symcache = SymCache::parse(&buffer)?;
    insta::assert_debug_snapshot_matches!("functions_linux", FunctionsDebug(&symcache));
    Ok(())
}

#[test]
fn test_load_header_macos() -> Result<(), Error> {
    let buffer = ByteView::open("../testutils/fixtures/symcache/current/macos.symc")?;
    let symcache = SymCache::parse(&buffer)?;
    insta::assert_debug_snapshot_matches!(symcache, @r###"SymCache {
    debug_id: DebugId {
        uuid: "67e9247c-814e-392b-a027-dbde6748fcbf",
        appendix: 0
    },
    arch: Amd64,
    has_line_info: true,
    has_file_info: true,
    functions: 1863
}"###);
    Ok(())
}

#[test]
fn test_load_functions_macos() -> Result<(), Error> {
    let buffer = ByteView::open("../testutils/fixtures/symcache/current/macos.symc")?;
    let symcache = SymCache::parse(&buffer)?;
    insta::assert_debug_snapshot_matches!("functions_macos", FunctionsDebug(&symcache));
    Ok(())
}

#[test]
fn test_lookup() -> Result<(), Error> {
    let buffer = ByteView::open("../testutils/fixtures/symcache/current/macos.symc")?;
    let symcache = SymCache::parse(&buffer)?;
    let line_infos = symcache.lookup(4_458_187_797 - 4_458_131_456)?;
    insta::assert_debug_snapshot_matches!("lookup", &line_infos);

    Ok(())
}
