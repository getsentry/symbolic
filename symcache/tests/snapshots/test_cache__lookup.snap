---
created: "2019-02-27T12:13:30.345704Z"
creator: insta@0.6.3
source: symcache/tests/test_cache.rs
expression: "&line_infos"
---
[
    LineInfo {
        arch: Amd64,
        debug_id: DebugId {
            uuid: "67e9247c-814e-392b-a027-dbde6748fcbf",
            appendix: 0
        },
        sym_addr: 56341,
        line_addr: 56341,
        instr_addr: 56341,
        line: 21,
        lang: Cpp,
        symbol: Some(
            "_ZN12_GLOBAL__N_15crashEv"
        ),
        filename: "main.cpp",
        base_dir: "../macos",
        comp_dir: "/Users/travis/build/getsentry/breakpad-tools/macos"
    },
    LineInfo {
        arch: Amd64,
        debug_id: DebugId {
            uuid: "67e9247c-814e-392b-a027-dbde6748fcbf",
            appendix: 0
        },
        sym_addr: 56341,
        line_addr: 56341,
        instr_addr: 56341,
        line: 25,
        lang: Cpp,
        symbol: Some(
            "_ZN12_GLOBAL__N_15startEv"
        ),
        filename: "main.cpp",
        base_dir: "../macos",
        comp_dir: "/Users/travis/build/getsentry/breakpad-tools/macos"
    },
    LineInfo {
        arch: Amd64,
        debug_id: DebugId {
            uuid: "67e9247c-814e-392b-a027-dbde6748fcbf",
            appendix: 0
        },
        sym_addr: 56224,
        line_addr: 56341,
        instr_addr: 56341,
        line: 32,
        lang: Cpp,
        symbol: Some(
            "main"
        ),
        filename: "main.cpp",
        base_dir: "../macos",
        comp_dir: "/Users/travis/build/getsentry/breakpad-tools/macos"
    }
]
